pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1247,7 +1250,18 @@ func (c *PodAffinityChecker) podMatchesPodAffinityTerms(pod, targetPod *v1.Pod,
 		if len(term.TopologyKey) == 0 {
 			return false, false, fmt.Errorf("empty topologyKey is not allowed except for PreferredDuringScheduling pod anti-affinity")
 		}
-		if !priorityutil.NodesHaveSameTopologyKey(nodeInfo.Node(), targetPodNode, term.TopologyKey) {
```

Comment:
In theory the operation is O(n), but in practice a node usually have ~10 label pairs, so I'm not quite a solution like this PR with O(1) implementation helps a lot. Your test result seems a little strange, I will run the same test again.

Reply:
@shiyan2016 testing in my local 32G-8cpu BM shows a different result:

```
## Without this PR

BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         131,230,945 ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         133,996,504 ns/op

BenchmarkScheduling/5000Nodes/1000Pods-8                  1000          89,063,795 ns/op
BenchmarkScheduling/5000Nodes/1000Pods-8                  1000          91,836,182 ns/op

## With this PR

BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         128,431,629 ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8                 1000         134,703,083 ns/op

BenchmarkScheduling/5000Nodes/1000Pods-8                  1000          93,030,224 ns/op
BenchmarkScheduling/5000Nodes/1000Pods-8                  1000          88,811,118 ns/op
```

Reply:
testing in my local 16G-4CPU shows there is not that much performance difference with or without this PR.

## Without this PR

```go
BenchmarkScheduling/5000Nodes/1000Pods-8                             1000	  44703105 ns/op
BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8    1000	  48068337 ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8           1000	  94545001 ns/op
BenchmarkSchedulingNodeAffinity/5000Nodes/1000Pods-8        1000	  45828025 ns/op
```

## With this PR

```go
BenchmarkScheduling/5000Nodes/1000Pods-8                           1000	  46596568 ns/op
BenchmarkSchedulingPodAntiAffinity/5000Nodes/1000Pods-8  1000	  46290482 ns/op
BenchmarkSchedulingPodAffinity/5000Nodes/1000Pods-8         1000	  95529805 ns/op
BenchmarkSchedulingNodeAffinity/5000Nodes/1000Pods-8       1000	  45912117 ns/op
```

Reply:
> In theory the operation is O(n), but in practice a node usually have ~10 label pairs, so I'm not quite a solution like this PR with O(1) implementation helps a lot. Your test result seems a little strange, I will run the same test again.

Replace NodesHaveSameTopologyKey using nodecache is mainly used to reduce node iteration and topology pair match check in InterPodAffinity processTerm: https://github.com/kubernetes/kubernetes/pull/79360/files#diff-ac123860716bbbb4ed0f26cc27062bb7R99.

Reply:
> > In theory the operation is O(n), but in practice a node usually have ~10 label pairs, so I'm not quite a solution like this PR with O(1) implementation helps a lot. Your test result seems a little strange, I will run the same test again.
> 
> Replace NodesHaveSameTopologyKey using nodecache is mainly used to reduce node iteration and topology pair match check in InterPodAffinity processTerm: https://github.com/kubernetes/kubernetes/pull/79360/files#diff-ac123860716bbbb4ed0f26cc27062bb7R99.

I'm quite curious about how did you run the benchmark with 2x performance improvement. Why did we get a different result on different machines?

Reply:
I also confused for the results on 2c6g. Maybe the cpu cores is the factor.